### PR TITLE
xds ring hash

### PR DESCRIPTION
**Pull request by sanxun0325**
**What this PR does**: 

**Which issue(s) this PR fixes**: 
Fixes #

**You should pay attention to items below to ensure your pr passes our ci test**
We do not merge pr with ci tests failed

- [x] All ut passed (run 'go test ./...' in project root)
- [x] After go-fmt ed , run 'go fmt project' using goland.
- [x] Golangci-lint passed, run 'sudo golangci-lint run' in project root.
- [x] After import formatted, (using [imports-formatter](https://github.com/dubbogo/tools#5-how-to-get-imports-formatter) to run 'imports-formatter .' in project root, to format your import blocks, mentioned in [CONTRIBUTING.md](https://github.com/apache/dubbo-go/blob/master/CONTRIBUTING.md) above) 
- [x] Your new-created file needs to have [apache license](https://raw.githubusercontent.com/dubbogo/resources/master/tools/license/license.txt) at the top, like other existed file does.
- [x] All integration test passed. You can run integration test locally (with docker env). Clone our [dubbo-go-samples](https://github.com/apache/dubbo-go-samples) project and replace the go.mod to your dubbo-go, and run 'sudo sh start_integration_test.sh' at root of samples project root. (M1 Slice is not Support)

### CODE REVIEW

File: cluster/loadbalance/xds/ringhash/ring.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	if i == len(items) {
+		// If not found, and h is greater than the largest hash, return the
+		// first item.
+		i = 0
+	}
+	return items[i]
+}
+
+func (lb *ringhashLoadBalance) generateHash(invocation protocol.Invocation, hashPolicies []*resource.HashPolicy) uint64 {
+	var hash uint64
```
Comment (AlexStocks) [inline cluster/loadbalance/xds/ringhash/ring.go:127]: 放在一个 var () 里面

**Comment by AlexStocks** [comment]
@sanxun0325 整体我觉得没啥问题，注意下 ut 覆盖率。

File: cluster/loadbalance/xds/ringhash/ring.go (hunk 1 above)
Comment (sanxun0325) [reply cluster/loadbalance/xds/ringhash/ring.go]: done

File: cluster/loadbalance/xds/ringhash/ringhash.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (LaurenceLiZhixin) [inline cluster/loadbalance/xds/ringhash/ringhash.go:1]: 这个文件能不能放到 cluster/loadbalance/下面

Comment (sanxun0325) [reply cluster/loadbalance/xds/ringhash/ringhash.go:1]: 是可以的，不过后面xds应该怀有其他的负载均衡能力，我想的是尽量把xds负载均衡的功能都搞在一起不和原来的混合

Comment (LaurenceLiZhixin) [reply cluster/loadbalance/xds/ringhash/ringhash.go:1]: 建议还是放到cluster/loadbalance/ringhash 下面吧。
负载均衡应该跟xds是解耦合的。

Comment (sanxun0325) [reply cluster/loadbalance/xds/ringhash/ringhash.go]: done

**Review by LaurenceLiZhixin — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**